modules/scala/almond-spark/src/main/scala/org/apache/spark/sql/almondinternals/ProgressSparkListener.scala

```diff
@@ -66,30 +69,77 @@ final class ProgressSparkListener(
         )
         elem.init(commTargetName, !sentInitCode)
         sentInitCode = true
+
+        updater.asyncPollUpdatesFor(elem)
       }
 
   override def onStageCompleted(stageCompleted: SparkListenerStageCompleted): Unit =
     if (progress)
       Try {
         val elem = stageElem(stageCompleted.stageInfo.stageId)
         elem.allDone()
-        elem.update()
       }
 
   override def onTaskStart(taskStart: SparkListenerTaskStart): Unit =
     if (progress)
       Try {
         val elem = stageElem(taskStart.stageId)
         elem.taskStart()
-        elem.update()
       }
 
   override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit =
     if (progress)
       Try {
         val elem = stageElem(taskEnd.stageId)
         elem.taskDone()
-        elem.update()
       }
 
 }
+
+private[almondinternals] class ProgressBarUpdater(
+  implicit publish: OutputHandler
+) {
+  private val threadFactory: ThreadFactory = {
+    val threadNumber = new AtomicInteger(1)
+    (r: Runnable) => {
+      val threadNumber0 = threadNumber.getAndIncrement()
+      val t = new Thread(r, s"almond-spark-listener-update-$threadNumber0")
+      t.setDaemon(true)
+      t.setPriority(Thread.NORM_PRIORITY)
+      t
+    }
+  }
+
+  private val pool = {
+    val executor = new ScheduledThreadPoolExecutor(1, threadFactory)
+    executor.setKeepAliveTime(1L, TimeUnit.MINUTES)
+    executor.allowCoreThreadTimeOut(true)
+    executor
+  }
+
+  private val pollings = new ConcurrentHashMap[StageElem, ScheduledFuture[_]]()
+
+  def asyncPollUpdatesFor(elem: StageElem): Unit = {
+    val polling = pool.scheduleAtFixedRate(
+      () => elem.update(),
+      0,
+      1,
+      TimeUnit.SECONDS
+    )
+    pollings.put(elem, polling)
+  }
+
+  pool.scheduleAtFixedRate(
```

Comment:
Do you think it would be possible to get rid of that one? In `() => elem.update()` above, couldn't `polling` be cancelled when `elem.allDone0` is true? My guess is that by typing `polling: ScheduledFuture[_]`, we can refer to it from the `Runnable`, and cancel it from there.

A benefit of that approach would be that the thread in the thread pool would get effectively shut down when no spark job is running, instead of being woken up every 2 seconds.

Reply:
Hmm, we only get the handle `polling` after submitting the runnable so I don't see how it would be possible to reference that in the runnable itself.
Another way to do it is to throw an excpetion when `elem.allDone0` to kind of force the runnable's termination.

Reply:
A `lazy val` for `polling` makes that work it seems, see the commit I just pushed.